File: src/Sample/Plumix.Sample/Demos/Material/NavigationDrawerDemoPage.cs
using System;
using Avalonia;
using Avalonia.Media;
using Plumix.Material;
using Plumix.Rendering;
using Plumix.UI;
using Plumix.Widgets;

namespace Plumix;

// Dart parity source: dart_sample/lib/demos/material/navigation_drawer_demo_page.dart

public sealed class NavigationDrawerDemoPage : StatefulWidget
{
    public override State CreateState() => new NavigationDrawerDemoPageState();

    private sealed class NavigationDrawerDemoPageState : State
    {
        private int? _selectedIndex;
        private bool _thirdEnabled = true;
        private bool _useThemeOverrides;
        private bool _useWidgetOverrides;

        public override Widget Build(BuildContext context)
        {
            var pageTheme = Theme.Of(context) with
            {
                NavigationDrawerTheme = _useThemeOverrides
                    ? new NavigationDrawerThemeData(
                        BackgroundColor: Color.Parse("#FFF3E5F5"),
                        IndicatorColor: Color.Parse("#FFB2DFDB"),
                        TileHeight: 60,
                        IndicatorSize: new Size(270, 48),
                        LabelTextStyle: MaterialStateProperty<TextStyle?>.All(
                            new TextStyle(Color: Color.Parse("#FF4A148C"), FontSize: 13)),
                        IconTheme: MaterialStateProperty<IconThemeData?>.All(
                            new IconThemeData(Color: Color.Parse("#FF00695C"), Size: 22)))
                    : new NavigationDrawerThemeData(),
            };

            return new Theme(
                data: pageTheme,
                child: new Column(
                    crossAxisAlignment: CrossAxisAlignment.Stretch,
                    spacing: 10,
                    children:
                    [
                        new Text("NavigationDrawer + NavigationDrawerDestination", fontSize: 20),
                        new Text(
                            "Header/footer slots, custom children, destination indexing, selection, disabled state, and theme precedence.",
                            fontSize: 14,
                            color: Colors.DimGray),
                        new Row(
                            spacing: 8,
                            children:
                            [
                                ControlButton(
                                    _useThemeOverrides ? "Theme on" : "Theme off",
                                    () => SetState(() => _useThemeOverrides = !_useThemeOverrides)),
                                ControlButton(
                                    _useWidgetOverrides ? "Widget on" : "Widget off",
                                    () => SetState(() => _useWidgetOverrides = !_useWidgetOverrides)),
                                ControlButton(
                                    _thirdEnabled ? "Disable third" : "Enable third",
                                    () => SetState(() => _thirdEnabled = !_thirdEnabled)),
                                ControlButton(
                                    _selectedIndex.HasValue ? "Clear selection" : "Select first",
                                    () => SetState(() => _selectedIndex = _selectedIndex.HasValue ? null : 0)),
                            ]),
                        new Text(
                            _selectedIndex.HasValue
                                ? $"Selected destination: {_selectedIndex.Value + 1}"
                                : "Selected destination: none",
                            fontSize: 13),
                        new Expanded(
                            child: new Row(
                                crossAxisAlignment: CrossAxisAlignment.Stretch,
                                spacing: 16,
                                children:
                                [
                                    new NavigationDrawer(
                                        selectedIndex: _selectedIndex,
                                        onDestinationSelected: index => SetState(() => _selectedIndex = index),
                                        backgroundColor: _useWidgetOverrides ? Color.Parse("#FFFFF8E1") : null,
                                        indicatorColor: _useWidgetOverrides ? Color.Parse("#FFFFCC80") : null,
                                        tilePadding: _useWidgetOverrides ? new Thickness(18, 0) : null,
                                        header: new Padding(
                                            new Thickness(28, 20, 16, 12),
                                            new Text("Destinations", fontSize: 16)),
                                        footer: new Padding(
                                            new Thickness(28, 12, 16, 20),
                                            new Text("Navigation footer", fontSize: 12, color: Colors.DimGray)),
                                        children:
                                        [
                                            new Padding(
                                                new Thickness(28, 8),
                                                new Text("Primary", fontSize: 12, color: Colors.DimGray)),
                                            new NavigationDrawerDestination(
                                                icon: new Icon(Icons.StarOutline),
                                                selectedIcon: new Icon(Icons.Star),
                                                label: new Text("Favorites")),
                                            new NavigationDrawerDestination(
                                                icon: new Icon(Icons.InfoOutline),
                                                label: new Text("Explore")),
                                            new Divider(indent: 28, endIndent: 28),
                                            new NavigationDrawerDestination(
                                                icon: new Icon(Icons.Menu),
                                                label: new Text("Downloads"),
                                                enabled: _thirdEnabled),
                                        ]),
                                    new Expanded(
                                        child: new Container(
                                            alignment: Alignment.Center,
                                            decoration: new BoxDecoration(
                                                Color: Color.Parse("#FFF7F2FA"),
                                                BorderRadius: BorderRadius.Circular(12)),
                                            child: new Text(
                                                "The drawer keeps destination indices independent from custom children.",
                                                color: Colors.DimGray,
                                                textAlign: TextAlign.Center)))
                                ]))
                    ]));
        }

        private static Widget ControlButton(string label, Action onPressed)
        {
            return new TextButton(
                child: new Text(label, fontSize: 12),
                onPressed: onPressed);
        }
    }
}

File: dart_sample/lib/demos/material/navigation_drawer_demo_page.dart
import 'package:material_ui/material_ui.dart';

class NavigationDrawerDemoPage extends StatefulWidget {
  const NavigationDrawerDemoPage({super.key});

  @override
  State<NavigationDrawerDemoPage> createState() =>
      _NavigationDrawerDemoPageState();
}

class _NavigationDrawerDemoPageState extends State<NavigationDrawerDemoPage> {
  int? _selectedIndex;
  bool _thirdEnabled = true;
  bool _useThemeOverrides = false;
  bool _useWidgetOverrides = false;

  @override
  Widget build(BuildContext context) {
    final ThemeData ambientTheme = Theme.of(context);
    final ThemeData pageTheme = ambientTheme.copyWith(
      navigationDrawerTheme: _useThemeOverrides
          ? const NavigationDrawerThemeData(
              backgroundColor: Color(0xFFF3E5F5),
              indicatorColor: Color(0xFFB2DFDB),
              tileHeight: 60,
              indicatorSize: Size(270, 48),
              labelTextStyle: WidgetStatePropertyAll<TextStyle?>(
                TextStyle(color: Color(0xFF4A148C), fontSize: 13),
              ),
              iconTheme: WidgetStatePropertyAll<IconThemeData?>(
                IconThemeData(color: Color(0xFF00695C), size: 22),
              ),
            )
          : const NavigationDrawerThemeData(),
    );

    return Theme(
      data: pageTheme,
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        spacing: 10,
        children: <Widget>[
          const Text(
            'NavigationDrawer + NavigationDrawerDestination',
            style: TextStyle(fontSize: 20),
          ),
          const Text(
            'Header/footer slots, custom children, destination indexing, selection, disabled state, and theme precedence.',
            style: TextStyle(fontSize: 14, color: Colors.black54),
          ),
          Wrap(
            spacing: 8,
            runSpacing: 8,
            children: <Widget>[
              _controlButton(
                _useThemeOverrides ? 'Theme on' : 'Theme off',
                () => setState(() => _useThemeOverrides = !_useThemeOverrides),
              ),
              _controlButton(
                _useWidgetOverrides ? 'Widget on' : 'Widget off',
                () =>
                    setState(() => _useWidgetOverrides = !_useWidgetOverrides),
              ),
              _controlButton(
                _thirdEnabled ? 'Disable third' : 'Enable third',
                () => setState(() => _thirdEnabled = !_thirdEnabled),
              ),
              _controlButton(
                _selectedIndex != null ? 'Clear selection' : 'Select first',
                () => setState(
                  () => _selectedIndex = _selectedIndex != null ? null : 0,
                ),
              ),
            ],
          ),
          Text(
            _selectedIndex != null
                ? 'Selected destination: ${_selectedIndex! + 1}'
                : 'Selected destination: none',
          ),
          Expanded(
            child: Row(
              crossAxisAlignment: CrossAxisAlignment.stretch,
              spacing: 16,
              children: <Widget>[
                NavigationDrawer(
                  selectedIndex: _selectedIndex,
                  onDestinationSelected: (int index) =>
                      setState(() => _selectedIndex = index),
                  backgroundColor: _useWidgetOverrides
                      ? const Color(0xFFFFF8E1)
                      : null,
                  indicatorColor: _useWidgetOverrides
                      ? const Color(0xFFFFCC80)
                      : null,
                  tilePadding: EdgeInsets.symmetric(
                    horizontal: _useWidgetOverrides ? 18 : 12,
                  ),
                  header: const Padding(
                    padding: EdgeInsets.fromLTRB(28, 20, 16, 12),
                    child: Text('Destinations', style: TextStyle(fontSize: 16)),
                  ),
                  footer: const Padding(
                    padding: EdgeInsets.fromLTRB(28, 12, 16, 20),
                    child: Text(
                      'Navigation footer',
                      style: TextStyle(fontSize: 12, color: Colors.black54),
                    ),
                  ),
                  children: <Widget>[
                    const Padding(
                      padding: EdgeInsets.symmetric(
                        horizontal: 28,
                        vertical: 8,
                      ),
                      child: Text(
                        'Primary',
                        style: TextStyle(fontSize: 12, color: Colors.black54),
                      ),
                    ),
                    const NavigationDrawerDestination(
                      icon: Icon(Icons.star_outline),
                      selectedIcon: Icon(Icons.star),
                      label: Text('Favorites'),
                    ),
                    const NavigationDrawerDestination(
                      icon: Icon(Icons.info_outline),
                      label: Text('Explore'),
                    ),
                    const Divider(indent: 28, endIndent: 28),
                    NavigationDrawerDestination(
                      icon: const Icon(Icons.menu),
                      label: const Text('Downloads'),
                      enabled: _thirdEnabled,
                    ),
                  ],
                ),
                const Expanded(
                  child: DecoratedBox(
                    decoration: BoxDecoration(
                      color: Color(0xFFF7F2FA),
                      borderRadius: BorderRadius.all(Radius.circular(12)),
                    ),
                    child: Center(
                      child: Text(
                        'The drawer keeps destination indices independent from custom children.',
                        textAlign: TextAlign.center,
                        style: TextStyle(color: Colors.black54),
                      ),
                    ),
                  ),
                ),
              ],
            ),
          ),
        ],
      ),
    );
  }

  Widget _controlButton(String label, VoidCallback onPressed) {
    return TextButton(onPressed: onPressed, child: Text(label));
  }
}
